rust/connlib/shared/src/interface/linux.rs

```diff
@@ -0,0 +1,296 @@
+//! Virtual network interface
+
+use crate::{
+    callbacks::{Cidrv4, Cidrv6},
+    linux::{etc_resolv_conf, DnsControlMethod},
+};
+use anyhow::{anyhow, bail, Context as _, Result};
+use futures::TryStreamExt;
+use ip_network::{IpNetwork, Ipv4Network, Ipv6Network};
+use netlink_packet_route::route::{RouteProtocol, RouteScope};
+use netlink_packet_route::rule::RuleAction;
+use rtnetlink::{new_connection, Error::NetlinkError, Handle, RouteAddRequest, RuleAddRequest};
+use std::{
+    collections::HashSet,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr},
+};
+
+// TODO: De-dupe before merging
+const FIREZONE_MARK: u32 = 0xfd002021;
+const IFACE_NAME: &str = "tun-firezone";
+const DEFAULT_MTU: u32 = 1280;
+const FILE_ALREADY_EXISTS: i32 = -17;
+const FIREZONE_TABLE: u32 = 0x2021_fd00;
+
+/// For lack of a better name
+pub struct InterfaceManager {
```

Comment:
```suggestion
pub struct TunDeviceManager {
```

Eventually, I think we should structure things such that we obtain a `Device` (or perhaps renamed `TunDevice` at that point) from this component and set it on the `Tunnel`.